### PR TITLE
security: add permissions for accessing resources

### LINKED ISSUE #146: Rules for restricting access to resources by user, role and institution

### LINKED ISSUE #217: Permissions rules for accessing deposits

### PR DESCRIPTION
**Pull request by sebastiendeleze**
* Creates a factory for managing permissions.
* Adds serializer and loader for deposit resources.
* Adds a deny access permission.
* Adds specific permissions decisions for all resources.
* Adds queries factories for filtering records in lists.
* Adds permissions results when serializing a record.
* Changes path in deposit validation email.
* Guesses user's organisation when creating a user record.
* Guesses user's organisation when creating a document record.
* Removes redirection to documents records when accessing administration.
* Adds permissions checks for adding resources when retrieving logged user by API.
* Removes organisation from user and document schema when logged user is not superuser.
* Improves fixtures in tests.
* Closes #146.
* Closes #217.

## How to test
For each roles (superuser, admin, moderator, publisher, user):
1. Login as user with the current role.
2. Check the access of the different resources. It must correspond to permissions matrix in https://docs.google.com/spreadsheets/d/1GyFbkm3VDCrffbShaFsoMJnkGQYtw65sWbfW25J2Mxk/edit#gid=0.

### CODE REVIEW

File: tests/ui/documents/test_documents_tasks.py [hunk 1]
```diff
@@ -28,6 +28,8 @@
 def test_import_records(mock_record_by_identifier, app, document_json,
                         bucket_location):
     """Test import records."""
+    document_json['pid'] = '10000'
```
Comment (jma) [inline tests/ui/documents/test_documents_tasks.py:31]: is this value should be really hard coded?

File: tests/ui/documents/test_documents_views.py [hunk 2]
```diff
@@ -44,7 +44,8 @@ def test_search(app, client):
 
 def test_detail(app, client, document):
     """Test document detail page."""
-    assert client.get('/organisation/sonar/documents/10000').status_code == 200
+    assert client.get('/organisation/sonar/documents/{pid}'.format(
```
Comment (jma) [inline tests/ui/documents/test_documents_views.py:47]: please use `url_for` if you can

File: tests/ui/test_api.py [hunk 3]
```diff
@@ -98,8 +105,8 @@ def test_get_record_by_ref_link(app, document):
     """Test getting a record by a reference link."""
 
     record = DocumentRecord.get_record_by_ref_link(
-        'https://sonar.ch/api/documents/10000')
-    assert record['pid'] == '10000'
+        'https://sonar.ch/api/documents/{pid}'.format(pid=document['pid']))
```
Comment (jma) [inline tests/ui/test_api.py:108]: probably you can use HOST_NAME or similar or better an absolute `url_for`

File: tests/ui/test_permissions.py [hunk 4]
```diff
@@ -86,24 +87,76 @@ def test_has_superuser_access(app, client, user_without_role, superuser):
...
+    app.config.update(SONAR_APP_DISABLE_PERMISSION_CHECKS=False)
+    assert not record_delete_permission_factory(document).can()
+
+    login_user_via_view(client, email=superuser['email'], password='123456')
+    assert record_delete_permission_factory(document).can()
+
+
+def test_unknown_permission_factory(app, client, superuser, document):
+    """Test unknown permission factory"""
+    app.config.update(SONAR_APP_DISABLE_PERMISSION_CHECKS=True)
```
Comment (jma) [inline tests/ui/test_permissions.py:152]: what is the idea of `SONAR_APP_DISABLE_PERMISSION_CHECKS`? Only for debug?

File: sonar/modules/deposits/marshmallow/json.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    @pre_dump
+    def add_permissions(self, item):
+        """Add permissions to record.
+
+        :param item: Dict representing the record
+        :returns: Dict of modified record.
+        """
+        item['permissions'] = {
+            'read': DepositPermission.read(current_user, item),
```
[the remark this answers is not in the document]
Comment (sebastiendeleze) [reply sonar/modules/deposits/marshmallow/json.py:67]: I don't understand your comment, what do you expect?

File: sonar/modules/documents/permissions.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        view = request.args.get('view')
+
+        # Documents are accessess in public view, but eventually filtered later
+        # by organisation
+        if view:
+            return True
+
+        # Only for moderators users.
+        if (not current_user_record or not current_user_record.is_moderator or
```
[the remark this answers is not in the document]
Comment (sebastiendeleze) [reply sonar/modules/documents/permissions.py:47]: Yes, I think it's useful for admin views. Only at least moderators have access to documents list.

File: sonar/modules/users/marshmallow/json.py [hunk 7]
```diff
@@ -44,13 +46,47 @@ class UserMetadataSchemaV1(StrictKeysMixin):
     city = SanitizedUnicode()
     phone = SanitizedUnicode()
     organisation = fields.Dict()
-    roles = fields.List(SanitizedUnicode, required=True)
```
[the remark this answers is not in the document]
Comment (sebastiendeleze) [reply sonar/modules/users/marshmallow/json.py:47]: It is always required by JSON schema. 
If you edit your own account, the roles cannot be changed. If I let the constraint here, the form will never be saved..

File: tests/ui/test_permissions.py (hunk 4 above)
Comment (sebastiendeleze) [reply tests/ui/test_permissions.py:152]: Yes, it disable completely permission checks for example for testing API with external tool like Postman.

File: sonar/modules/users/api.py [hunk 8]
```diff
@@ -98,7 +98,9 @@ class UserRecord(SonarRecord):
     fetcher = user_pid_fetcher
     provider = UserProvider
     schema = 'users/user-v1.0.0.json'
-    available_roles = [ROLE_SUPERUSER, ROLE_ADMIN, ROLE_MODERATOR, ROLE_USER]
+    available_roles = [
```
[the remark this answers is not in the document]
Comment (sebastiendeleze) [reply sonar/modules/users/api.py:101]: Yes, I think it's possible, but It's not the scope of this PR. I will add a new issue for that.

File: sonar/modules/deposits/marshmallow/json.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jma) [reply sonar/modules/deposits/marshmallow/json.py:67]: computing the permissions for all search results can take while. Try to display 1000 results and check the response time.

File: tests/ui/test_permissions.py (hunk 4 above)
Comment (jma) [reply tests/ui/test_permissions.py:152]: ok

**Pull request merged**